File: aoc/aoc2022day04campcleanup2.cpp
// Vicfred
// https://adventofcode.com/2022/day/4
// implementation
#include <cstdio>
#include <fstream>
#include <iostream>

using namespace std;

int main() {
  ifstream input("input.txt");
  long long ans = 0LL;
  for (string line; getline(input, line);) {
    int v, w, x, y;
    sscanf(line.c_str(), "%d-%d,%d-%d", &v, &w, &x, &y);
    if (v == x) {
      ans += 1;
      continue;
    }
    if(v < x) {
      swap(v, x);
      swap(w, y);
    }
    if(y >= v) {
      ans += 1;
    }
  }
  cout << ans << endl;
  return 0;
}

File: aoc/aoc2022day04campcleanup2.hs
-- uninhm
-- https://adventofcode.com/2022/day/4#part2
-- implementation

{-# LANGUAGE OverloadedStrings #-}

import qualified Data.Text as T

data Range a = Range
  { start :: a
  , end :: a
  } deriving (Show, Eq)

fromList :: [a] -> Range a
fromList [s, e] = Range s e

overlapsWith :: Ord a => Range a -> Range a -> Bool
overlapsWith other range
  | start range > start other = overlapsWith range other
  | otherwise = end range >= start other

parse :: String -> [[Range Int]]
parse = map parseLine . T.lines . T.pack
  where
    parseLine line =
      fromList . map (read . T.unpack) . T.splitOn "-" <$> T.splitOn "," line

solve :: [[Range Int]] -> Int
solve = length . filter id . map checkOverlaping
  where checkOverlaping [a, b] = overlapsWith a b

main :: IO ()
main = print . solve . parse =<< readFile "input"
